리플렉션 실습 3

kangminhyuk1111 opened the pull request:
리플렉션 실습 3단계 중 생성자 부분만 먼저 해봤습니다. 
어제 말씀해주신 예외처리 부분이 조금 생각이 안나는데 제가 예외 처리를 안한 부분이 어디에 있을까요? 
일단 제가 생각했을 때는 getAutowiredConstructor 에서 autowired가 1개가 아닌 경우일때 예외처리 해줬습니다.

kangminhyuk1111 committed:
리플렉션 실습 3단계

리플렉션 실습 3단계 중 생성자 부분만 먼저 해봤습니다. 어제 말씀해주신 예외처리 부분이 조금 생각이 안나는데 제가 예외 처리를 안한 부분이 어디에 있을까요? 일단 제가 생각했을 때는 getAutowiredConstructor 에서 autowired가 1개가 아닌 경우일때 예외처리 해줬습니다.

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -28,7 +33,12 @@ public class ApplicationContext {
     // 빈에 등록될 클래스들을 모아놓은 리스트
     private final Set<Class<?>> beanClasses = new HashSet<>();
-    private Map<Class<?>, Object> beans = new HashMap<>();
+    // 생성되지 않은 빈들의 목록 String값으로 접근 가능
+    // getName() 으로 접근해서 클래스의 정보를 가져온 다음 인스턴스 생성하도록
+    private final Map<String, Object> beanMap = new HashMap<>();
+    // autowired에 필요한 인스턴스 생성
+    // beanClasses에 등록된 클래스들을 실제 인스턴스로 만듬
+    private final Map<Class<?>, Object> beans = new HashMap<>();
```
`beanMap` 과 `beans` 두개의 Map 이 필요한 이유를 잘 모르겠어요.
의존성 주입이 되어야 하는 객체들도 결국 bean 이기 때문에 한번에 관리해도 괜찮지 않을까요?

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -38,11 +48,83 @@ public ApplicationContext(final String packageName) {
         // 위에 찾은 클래스들을 beanClasses에 저장
         beanClasses.addAll(annotationClasses);
         // 빈을 동적으로 생성하는 메서드 실행
         createBeans(beanClasses);
+        createBean(beanClasses);
```
`createBeans` 와 `createBean` 두개의 메서드가 존재해야 하는지 잘 모르겠네요.
bean 을 생성하기 위한 메서드는 하나만 있어도 되지 않을까요?

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -38,11 +48,83 @@ public ApplicationContext(final String packageName) {
+        }
+    }
+
+    private boolean isContainsKey(Class<?> beanClass) {
+        return beanMap.containsKey(beanClass.getSimpleName());
+    }
```
미션과 별개로 이런 메서드 이름은 좋지 못한거 같아요.
이 메서드가 정확히 어떤 역할을 하는지 명시적으로 나타내는게 좋아요.

`isContainsKey` 라고만 하면 key 가 포함되어있는지 boolean 으로 반환해준다 는 역할일거 같은데
어디에 어떤 키가 포함되었는지, 포함되는지 여부가 왜 중요한지 의미를 파악하기 어려운거 같아요.

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -38,11 +48,83 @@ public ApplicationContext(final String packageName) {
+                parameters[i] = beanMap.get(parameterTypes[i].getSimpleName());
+            }
+
+            parameters[i] = parameterTypes[i].newInstance();
```
이런식으로 파라미터를 생성하는 메서드에서 개별적으로 생성해줄 때, 저희가 저번 단계에서 의도한 싱글톤 패턴을 잘 활용할 수 있을까요?

예를 들어 UserService 가 UserRepository 에 의존하고 있어서 생성자 주입이 필요하다고 할 때,
UserRepository 가 bean 으로 생성된 객체를 받기를 기대하는데 여기서 생성해버리면 파라미터로 주입되는 UserService 의 UserRepository 와 빈으로 생성된 UserRepository 가 다른 객체가 될거 같아요.

이런 부분을 검증하려면 테스트 코드를 작성하면서 개발하는것이 좋아요.

``` java
final UserService userService = getBean(UserService.class);
final UserRepository userRepository = getBean(UserRepository.class);
assertThat(userService.userRepository).isEqualTo(userRepository);
```

이런식으로 검증하면 싱글톤인지 체크할 수 있을거 같아요.

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -61,13 +143,8 @@ private void createBeans(Set<Class<?>> classes) {
-            Field[] declaredFields = annotationClass.getDeclaredFields();
-            Constructor<?>[] declaredConstructors = annotationClass.getDeclaredConstructors();
-        }
+    private Object saveBean(String beanName, Object instance) {
```
이 메서드의 반환값을 사용하는 코드가 존재하지 않아요!
메서드를 void 로 작성하는게 좋지 않을까요?

djawnstj reviewed (changes requested):
민혁님!
아직 미션의 감이 잘 안잡힐 수 있어요.
핵심은
1. 빈을 싱글톤으로 생성한다.
2. 생성중 의존성 주입이 필요한 빈을 만나면 의존성 주입을 해준다.
3. 의존성 주입을 해줄 때 빈이 만들어져있지 않으면 해당 빈은 새로 생성해준다(1번으로)
이 흐름만 알면 돼요.

그리고 미션을 진행하면서 기능이 추가되면 테스트를 작성하는게 어떨까요?
먼저 요구사항에 맞게 테스트 목록을 만들어 보고 미션을 진행하면서 내 의도에 맞는 코드가 개발되면
그 의도를 테스트하면서 제대로 작동되는지 확인해 보면 좋을거 같아요.

예외상황은 잘 만들었는데 추가로 파라미터 타입이 프레임워크가 관리할 빈인지도 체크하면 좋을거 같아요.
빈이 아닌데 Autowired 되면 안되기 때문이에요.

이런 예외 케이스들도 테스트로 검증을 해두면 좋겠죠??

진행중에 고민이 생기면 언제든 DM 주세요!
화이팅 입니다.

kangminhyuk1111 committed:
리플렉션 실습 3단계

피드백 반영해서 테스트 코드 통과하도록 한번 작성 해봤습니다.

어떤 부분을 수정하면 좋을까요?

이번에는 createAndSaveBean 메서드를 생성해서 없는 경우에 생성하고 있는 경우는 beanMap에서 찾아서 사용하도록 했고 테스트코드 통과까지 확인하였습니다.

kangminhyuk1111 commented:
리플렉션 실습 3단계
피드백 반영해서 테스트 코드 통과하도록 한번 작성 해봤습니다.

어떤 부분을 수정하면 좋을까요?

이번에는 createAndSaveBean 메서드를 생성해서 없는 경우에 생성하고 있는 경우는 beanMap에서 찾아서 사용하도록 했고 테스트코드 통과까지 확인하였습니다.

+ 주석에 대해서 고민을 좀 해봤는데 제가 이해하는데 도움이 되려고 주석달고 설명적으면서 코드 적은 것 같은데, 코드에 주석이 어느정도까지 있는게 좋다고 생각하시나요? 저번에 클린코드에서는 주석이 없는 코드가 이상적이라고 하신 것 같은데 또 아예 없는거는 가독성이 조금 떨어질 것 같다는 생각이 들고 너무 과하면 난잡해질 것 같아서 여쭤봅니다.

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -20,52 +20,33 @@ public class ApplicationContext {
     }
 
-    private boolean isContainsKey(Class<?> beanClass) {
+    private boolean isBeanExist(Class<?> beanClass) {
```
피드백 반영👍🏻

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -91,16 +72,27 @@ private Object[] createParameters(Class<?>[] parameterTypes) throws Instantiatio
         return parameters;
     }
 
+    private Object createAndSaveBean(Class<?> parameterType) throws InstantiationException, IllegalAccessException {
```
`createAndSaveBean` 메서드를 보자마자 '역할을 나누는게 좋지 않을까?' 생각을 했어요.
그런데 내부 구현 로직은 `saveBean()` 을 호출하는 방식이더라고요.

저는 이런 경우엔 메서드 명을 `createBean` 으로 할거 같은데 민혁님은 어떤가요?

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -91,16 +72,27 @@ private Object[] createParameters(Class<?>[] parameterTypes) throws Instantiatio
+            return beanMap.get(parameterType.getSimpleName());
+        }
+
+        Object instance = parameterType.newInstance();
+        saveBean(parameterType.getSimpleName(), instance);
```
일단 `Class.newInstance()` 보단 `Constructor.newInstance()` 로 인스턴스를 생성하는게 좋아요!
두번째로 의존성 주입이 제대로 안될거 같은 문제가 있어요.

만약 `UserController(UserService)`, `UserService(UserRepository)`, `UserRepository()` 관계의 빈을 만든다고 하면,
`UserController` 를 생성하기 위해 `UserService` 가 필요한데 지금같은 로직에선 `UserService` 를 생성할 때 `UserRepository` 를 주입해주지 못할거 같아요.

민혁님은 지금처럼 파라미터가 되는 빈 인스턴스를 생성하는 로직이 별도로 만들어 둔 이유가 있을까요?

djawnstj commented on src/test/java/study/context/ApplicationContextTest.java:
```diff
@@ -55,4 +55,19 @@ void addBeanSingletonTest() {
                 () -> Assertions.assertThat(bean1).isSameAs(bean2));
     }
+
+    @Test
+    @DisplayName("빈 등록 테스트 - 싱글톤 테스트")
+    void singletonTest() {
+        ApplicationContext applicationContext = new ApplicationContext("com.djawnstj.mvcframework");
+        applicationContext.init();
+
+        final UserService userService = applicationContext.getBean(UserService.class);
+        final UserRepository userRepository = applicationContext.getBean(UserRepository.class);
+
+        assertAll("foo",
+                () -> Assertions.assertThat(userService).isNotNull(),
+                () -> Assertions.assertThat(userRepository).isNotNull(),
+                () -> Assertions.assertThat(userService.userRepository).isEqualTo(userRepository));
+    }
```
저희가 저번에 본대로 주입되는 빈의 이름을 오름차순 기준으로 후순으로 가도록 변경해서도 정상적으로 주입되는지 확인해보면 좋을거 같아요.
위의 피드백에서 남긴대로 3단계 의존 관계를 대상으로 테스트를 진행해보면 더 좋을거 같아요!

djawnstj commented on src/test/java/study/context/ApplicationContextTest.java:
```diff
@@ -55,4 +55,19 @@ void addBeanSingletonTest() {
+        final UserService userService = applicationContext.getBean(UserService.class);
+        final UserRepository userRepository = applicationContext.getBean(UserRepository.class);
+
+        assertAll("foo",
```
"foo" 처럼 의미 없는 이름은 커밋 전에 수정해두는게 좋습니다!

djawnstj reviewed (changes requested):
민혁님!
수정해야 하는 부분 피드백 남겼어요.

몇가지 문제가 더 있는데요,
1. 테스트가 실행이 안됩니다. `PayServiceTest` 클래스에서 컴파일 에러가 발생하는데 제 환경이 문제인지 확인 부탁드립니다.
2. 테스트코드 패키지가 아직 이상해요ㅠㅠ 지난 단계 마지막 피드백으로 남겼는데 확인해서 수정 부탁드립니다.
3. 의존성 주입 시 파라미터 타입이 빈으로 생성해야 하는 대상인지 체크되고 있지 않은거 같아요.

위 세가지는 같이 수정해주면 좋을거 같아요.

주석에 대한 궁금증을 남겨주셨는데, 저는 기존에 적혀있던 주석은 전혀 읽어보지 않았습니다.
저한테는 오히려 코드가 연속적으로 보여야 하는데, 코드 중간중간에 주석이 있어서 가독성이 안좋아지더라고요😢 다만 민혁님이 학습하는 과정에서 정리의 용도로 적었다 생각해 따로 피드백은 안남겼습니다.

사람마다 다를 수 있긴 하지만 저는 주석이 아예 없는게 좋은거 같아요.
그렇기 때문에 원래 코드를 짤때 주석을 하나도 남기지 않지만 제 코드리뷰를 해주시는 분들께 주석에 대한 피드백은 단한번도 받지 못했어요.
저번에 말했듯이 클린코드에서도 여러 이유로 주석은 불필요하다고 말하고 있어요. 하지만 프레임워크나 라이브러리는 주석을 JavaDoc 으로 주석을 남기고 이용자가 코드의 역할을 분명히 알 수 있도록 남기는게 좋을 수 있다고 말하고 있긴 합니다. 그대신 관리를 꼼꼼히 해야갰죠?

만약 민혁님이 타인의 코드를 읽을 때 주석이 없어서 코드를 이해하는데 어려움이 발생한다면
저는 '주석을 달아주세요' 라고 말하는거 보단 주석 없이 코드를 파악하는 연습을 하는게 좋다고 생각합니다!

kangminhyuk1111 committed:
local

kangminhyuk1111 committed:
리플렉션 실습 3 - 피드백 수정

패키지 변경

kangminhyuk1111 committed:
리플렉션 실습 3 - 피드백 수정

UserController 추가해서 테스트 케이스 하나 추가해봤고 통과했습니다.

kangminhyuk1111 committed:
local

kangminhyuk1111 committed:
local

kangminhyuk1111 committed:
리플렉션 실습 3

지금 문제가 configuration에서 payservice를 생성할 때, payRepository가 생성이 되어 있어야하는데 없어서 문제가 생기는 것 같습니다.

3-2를 구현하는 방법은
1. configuration 안에 bean들을 모두 map으로 저장시킨후,

2. configuration의 내부의 메서드들을 실행시킬때 매개변수로 들어있는 빈을 찾고

3. 만약 매개변수로 들어있는 빈을 생성시켜서 saveBean으로 저장시키고 사용한다.

4. 만약에, beanClasses에도 없고, configuration에도 존재하지 않으면 exception

예를들어 payService를 생성할 때, 매개 변수인 payRepository가 있어야되기 때문에 @Bean으로 어노테이션이 되어있는지 찾아보고 있으면 그 빈을 인스턴스로 생성해서 주입,
만약에 @Bean으로 없으면 인스턴스가 생성되어서 저장되어있는 beanMap에서 찾을 것 같습니다. 둘다없으면 에러가 날 것 같고요.

그래서일단 returnType을 key로 가지는 map을 선언하고 configurationMap.get(returnType) 했을때 존재하면 configuration에 존재하니까 인스턴스로 생성해서 bean에 저장하고 매개변수로 넣어주면 될 것 같습니다.

말로는 되는데 실제로 코드 작성하려니까 잘 안되서 어떻게 하면 좋을지 여쭤봅니다

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -1,58 +1,60 @@
+    private void createBeansReferenceByConfiguration(final Set<Class<?>> annotationClasses) {
+        for (Class<?> annotationClass : annotationClasses) {
+            if (annotationClass.isAnnotationPresent(Configuration.class)) {
                 try {
-                    Object instance = beanClass.getDeclaredConstructor().newInstance();
-                    // {클래스 : 인스턴스} 로 hashmap에 저장
-                    beans.put(beanClass, instance);
+                    Object instance = annotationClass.getDeclaredConstructor().newInstance();
+                    Method[] methods = annotationClass.getDeclaredMethods();
+
+                    for (Method method : methods) {
+                        if (method.isAnnotationPresent(Bean.class)) {
+                            // 리턴 타입, 메소드
+                            configurationMap.put(method.getReturnType(),method);
+                        }
+                    }
                 } catch (InstantiationException | IllegalAccessException | InvocationTargetException |
                          NoSuchMethodException e) {
                     throw new RuntimeException(e);
```
이 메서드에서 instance 변수는 더 이상 사용되지 않아요.
특히 `newInstance()` 메서드의 경우 예외를 발생시킬 수 있기 때문에 사용하지 않다면 바로 지워주는게 좋을 것 같습니다!

그리고 중첩된 for, if 문이 너무 많은 것 같아요😭
저희 스터디 과정에 나오는 [📖 객체지향설계](https://djawnstj.notion.site/6f4c61dbca33411e8c6fb1312a2f95fb?pvs=4) 문서에 보면
> ### **객체지향 생활 체조 원칙 (참고만)**
> - 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기만 한다.
> ...

가 나옵니다. 

이런 규칙을 잘 지키면 더 좋은 코드로 변할 것 같아요!

djawnstj reviewed (changes requested):
민혁님!
기존 빈 생성 기능과 함께 `Configuration` 의 `@Bean` 기능을 추가하다 보니 어려운 점이 많을 거에요.

제 생각은 '굳이 beanClasses 와 configurationMap 이 나눠져야 하나?' 입니다.
`@Bean` 이든 `@Component` 든 결국 생성되어야 하는 빈 인건 마찬가지 인데요.

이런 접근 방식이라면 '`@Bean` 을 먼저 생성하자!` 또는 '`@Component` 를 먼저 생성하자! 의 접근에서 벗어날 수 있을 것 같습니다.

제 피드백이 너무 추상적이라면 말씀 주세요! 같이 고민해 봐도 좋을 것 같습니다.

kangminhyuk1111 committed:
리플렉션 실습 3-2

kangminhyuk1111 committed:
리플렉션 실습 3-2 수정

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -95,23 +101,39 @@ private void createInstance(final Class<?> beanClass) {
-            // 클래스에 없거나, config에도 없거나 둘다 없을때
-            if (!(beanClasses.contains(parameterType)) && !(configurationMap.containsKey(parameterType))) {
+            // Set에 없는경우
+            if (!(beanClasses.contains(parameterType))) {
```
하나의 메서드에는 하나의 들여쓰기만 존재한다! 원칙을 되새기면서 작업 하면 더 좋을 것 같습니다!

djawnstj commented on src/main/java/com/djawnstj/mvcframework/context/ApplicationContext.java:
```diff
@@ -20,45 +18,39 @@
     private final Set<Class<?>> beanClasses = new HashSet<>();
     public final Map<String, Object> beanMap = new HashMap<>();
-    public final Map<Class<?>, Object> configurationMap = new HashMap<>();
+    public final Map<String, Method> configurationMap = new HashMap<>();
```
이 두 필드는 public 일 이유가 없을 것 같아요!

djawnstj reviewed (approved):
민혁님!! 잘 해주셨습니다.
추가 피드백 조금 남겼는데 다음 미션 때 같이 반영 해주시면 좋을 것 같아요.
수고하셨습니다!!
